OpenStack Firewall as a Service

ggiamarchi opened the pull request:
**Prerequisites**
- [x] Need PR https://github.com/rackspace/gophercloud/pull/342 to be merged first

**To-do remaining**
- [x] Refactor a bit for code style consistency
- [x] Fix corner case bugs that fail some acceptance tests

ggiamarchi committed:
Add FWaaS rule resource

ggiamarchi committed:
Add FWaaS policy resource

ggiamarchi committed:
Add FWaaS firewall resource

ggiamarchi committed:
Add FWaaS rule acceptance test

ggiamarchi committed:
Enable FWaaS resources

ggiamarchi committed:
Add FWaaS policy acceptance test

ggiamarchi committed:
Add FWaaS firewall acceptance test

ggiamarchi committed:
Add tenant_id attribute on FWaaS resources

ggiamarchi committed:
Fix code regarding to the latest gophercloud code

ggiamarchi committed:
Fix race conditions on firewall state transition

ggiamarchi committed:
Remove useless code

ggiamarchi committed:
Fix firewall policies tests

ggiamarchi committed:
Unassociate firewall rule from policy before delete

ggiamarchi committed:
Use d.GetOk to populate data in read operations

ggiamarchi commented:
Work is done here

jrperritt commented:
Cool. I'll try to find some time tomorrow to review.

jrperritt commented on builtin/providers/openstack/resource_openstack_fw_firewall_v2.go:
```diff
@@ -0,0 +1,270 @@
+
+func resourceFWFirewallV2() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceFirewallCreate,
```
We'll want to add versions to these function names like in the other resource files.

ggiamarchi replied:
Done bd55765d7a3364cfcc3b04311cee119a96363419.

jrperritt commented on builtin/providers/openstack/resource_openstack_fw_firewall_v2.go:
```diff
@@ -0,0 +1,270 @@
+	}
+
+	firewall, err := firewalls.Get(networkingClient, d.Id()).Extract()
+	if err != nil {
```
Can we use `CheckDeleted` here? It's in `util.go`.

ggiamarchi replied:
Done 395016b78f36800d320f5ad81f73a9ce6d6785b2

jrperritt commented on builtin/providers/openstack/resource_openstack_fw_firewall_v2.go:
```diff
@@ -0,0 +1,270 @@
+
+	_, err = stateConf.WaitForState()
+
+	return firewalls.Update(networkingClient, d.Id(), opts).Err
```
This line should be before `stateConf` is defined. And up there it should be:

```
err  = firewalls.Update(networkingClient, d.Id(), opts).Err
if err != nil {
    // ....
}
```

Then, the return statement should be something like:
`return resourceFirewallV2Read()`

jrperritt replied:
Oh, I see. You're waiting for a pending create or update request to finish before calling this update request. Then perhaps just move the `Update` function one lineup and add the return statement above.

ggiamarchi replied:
Done 3321c74886ef75541e9ca5ee200b220bb639b716

jrperritt commented on builtin/providers/openstack/resource_openstack_fw_firewall_v2.go:
```diff
@@ -0,0 +1,270 @@
+	log.Printf("[DEBUG] Updating firewall with id %s: %#v", d.Id(), opts)
+
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{"PENDING_CREATE", "PENDING_UPDATE"},
```
Is `PENDING_CREATE` a valid state when updating? Or is this for when a user makes an update request while the firewall is still being created?

ggiamarchi replied:
This is exactly for when a user makes an update request while the firewall is still being created.

jrperritt commented on builtin/providers/openstack/resource_openstack_fw_policy_v2.go:
```diff
@@ -0,0 +1,231 @@
+
+func resourceFirewallPolicyCreate(d *schema.ResourceData, meta interface{}) error {
+
+	// TODO To remove
```
Should this still be here?

ggiamarchi replied:
Oups ! It was just for debugging. Deleted 2a5ff8d5da2ee2bb0f169bb0e13a1283d8bbffcb

jrperritt commented on builtin/providers/openstack/resource_openstack_fw_rule_v2.go:
```diff
@@ -0,0 +1,266 @@
+
+	rule, err := rules.Get(networkingClient, d.Id()).Extract()
+
+	if err != nil {
```
`CheckDeleted`

jrperritt commented:
Just a few comments from me. I can't run the tests since I don't have FWaaS, but hopefully when this gets merged others will use it and report any issues. Look good to me though.

ggiamarchi committed:
FWaaS version is actually v1 not v2

Rename files and methods.

Confusion have been made between neutron
version and FWaaS extension version.

ggiamarchi committed:
remove boilerplate code using CheckDeleted

ggiamarchi committed:
Return Read call result in Create & Update

ggiamarchi committed:
Code clean-up

ggiamarchi commented:
@julienvey Can you run acceptance tests and validate the behaviour ?

ggiamarchi commented:
@jrperritt Fixes about your comments are done.

julienvey commented:
Acceptance tests are sucessful 

Here is the full log http://paste.openstack.org/show/178228/

jrperritt commented:
Thank you for running and confirming the tests, @julienvey .
LGTM :rocket:
